x/reward/keeper/reward_claim.go

```diff
@@ -0,0 +1,170 @@
+	}
+	if state.GetClaimStatus() != types.RewardAccountState_CLAIM_STATUS_CLAIMABLE {
+		return reward, false
+	}
+
+	distribution, err := k.GetClaimPeriodRewardDistribution(ctx, period, rewardProgram.GetId())
+	if err != nil {
+		return reward, false
+	}
+
+	participantReward := k.CalculateParticipantReward(ctx, int64(state.GetSharesEarned()), distribution.GetTotalShares(), distribution.GetRewardsPool(), rewardProgram.MaxRewardByAddress)
+	reward = types.ClaimedRewardPeriodDetail{
+		ClaimPeriodId:     period,
+		TotalShares:       state.GetSharesEarned(),
+		ClaimPeriodReward: participantReward,
+	}
+
+	state.ClaimStatus = types.RewardAccountState_CLAIM_STATUS_CLAIMED
+	k.SetRewardAccountState(ctx, state)
+
+	return reward, true
+}
+
+// sendRewards internal method called with ClaimedRewardPeriodDetail of a single reward program
+func (k Keeper) sendRewards(ctx sdk.Context, rewards []*types.ClaimedRewardPeriodDetail, addr string, rewardProgramDenom string) (sdk.Coin, error) {
+	amount := sdk.NewInt64Coin(rewardProgramDenom, 0)
+
+	if len(rewards) == 0 {
+		return amount, nil
+	}
+
+	for _, reward := range rewards {
+		amount.Denom = reward.GetClaimPeriodReward().Denom
+		amount = amount.Add(reward.GetClaimPeriodReward())
+	}
+
+	return k.sendCoinsToAccount(ctx, amount, addr)
+}
+
+// sendCoinsToAccount internal wrapper method, to mainly do `SendCoinsFromModuleToAccount`
+func (k Keeper) sendCoinsToAccount(ctx sdk.Context, amount sdk.Coin, addr string) (sdk.Coin, error) {
+	if amount.IsZero() {
+		return sdk.NewInt64Coin(amount.GetDenom(), 0), nil
+	}
+
+	acc, err := sdk.AccAddressFromBech32(addr)
+	if err != nil {
+		return sdk.NewInt64Coin(amount.Denom, 0), err
+	}
+
+	err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, acc, sdk.NewCoins(amount))
+	if err != nil {
+		return sdk.NewInt64Coin(amount.Denom, 0), err
+	}
+
+	return amount, nil
+}
+
+// RefundRewardClaims refund all unclaimed rewards to the reward program creator
+func (k Keeper) RefundRewardClaims(ctx sdk.Context, rewardProgram types.RewardProgram) error {
+	amount := rewardProgram.TotalRewardPool.Sub(rewardProgram.RemainingPoolBalance).Sub(rewardProgram.ClaimedAmount)
+	_, err := k.sendCoinsToAccount(ctx, amount, rewardProgram.GetDistributeFromAddress())
+	return err
+}
+
+// ClaimAllRewards calls ClaimRewards, however differs from ClaimRewards in that it claims all the rewards that the address
+// is eligible for across all reward programs.
+func (k Keeper) ClaimAllRewards(ctx sdk.Context, addr string) ([]*types.RewardProgramClaimDetail, sdk.Coins, error) {
+	var allProgramDetails []*types.RewardProgramClaimDetail
+	allRewards := sdk.Coins{}
+	err := k.IterateRewardPrograms(ctx, func(rewardProgram types.RewardProgram) (stop bool, err error) {
+		// ignore expired reward programs from all claim( i.e do not error on them)
+		if rewardProgram.State != types.RewardProgram_STATE_EXPIRED {
+			details, reward, err := k.ClaimRewards(ctx, rewardProgram.GetId(), addr)
```

Comment:
It's generally bad practice to update state stuff while you're iterating over it. Weird things can happen to the iterator. A lot of times you can get away with it, specially when state is small, but it'll come back to bite you later.

Reply:
Agreed, the documentation even mentions this.